lib/web/terminal.go

```diff
@@ -217,15 +222,50 @@ func (t *TerminalHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		CheckOrigin:     func(r *http.Request) bool { return true },
 	}
 
+	emitError := func(errMsg string, err error) {
+		t.log.Errorf("%s: %s", errMsg, err)
+		http.Error(w, errMsg, http.StatusInternalServerError)
+	}
+
 	ws, err := upgrader.Upgrade(w, r, nil)
 	if err != nil {
-		errMsg := "Error upgrading to websocket"
-		t.log.Errorf("%v: %v", errMsg, err)
-		http.Error(w, errMsg, http.StatusInternalServerError)
+		emitError("Error upgrading to websocket", err)
+		return
+	}
+
+	ws.SetReadDeadline(deadlineForInterval(t.keepAliveInterval))
+
+	// If the displayLogin is set then use it instead of the login name used in
+	// the SSH connection. This is specifically for the use case when joining
+	// a session to avoid displaying "-teleport-internal-join" as the username.
+	if t.displayLogin != "" {
+		t.sessionData.Login = t.displayLogin
+	}
+
+	sessionString, err := json.Marshal(siteSessionGenerateResponse{Session: t.sessionData})
+	if err != nil {
+		emitError("unable to marshal session response", err)
```

Comment:
IIRC you can't write into the response writer here since the connection is upgraded to WebSockets and it should return something similar to: `response.Write on hijacked connection`

maybe it's something we should do is introduce a new protocol negotiation where we can have an error and payload flows. Since go does not support WebSockets over h2, we can do something like defining the stream id:
```
stdout = 0x1
error = 0x2
```
And once you write into the WebSocket you select the stream id. the server will prefix the messages with the streamID selected and once the client receives a message in the error stream it just exits.

Reply:
These errors are now being sent over the websocket as an error type message.